torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -217,22 +218,51 @@ void TensorPipeAgent::startImpl() {
   std::string lowestPriorityTransport;
 
   for (auto& key : TensorPipeTransportRegistry()->Keys()) {
+    int64_t priority = -1;
+    if (opts_.transports.has_value()) {
+      auto iter =
+          std::find(opts_.transports->begin(), opts_.transports->end(), key);
+      if (iter == opts_.transports->end()) {
+        continue;
+      }
+      // Assign priorities in reverse order of occurrence in the vector, so that
+      // a transport that comes before another receives a higher priority.
+      priority =
+          opts_.transports->size() - 1 - (iter - opts_.transports->begin());
+    }
     std::unique_ptr<TransportRegistration> reg =
         TensorPipeTransportRegistry()->Create(key);
-    if (reg->priority < lowestPriority) {
-      lowestPriority = reg->priority;
+    if (priority == -1) {
+      priority = reg->priority;
+    }
+    if (priority < lowestPriority) {
+      lowestPriority = priority;
       lowestPriorityTransport = key;
     }
     addresses.push_back(c10::str(key, "://", reg->address));
     context_->registerTransport(
-        reg->priority, std::move(key), std::move(reg->transport));
+        priority, std::move(key), std::move(reg->transport));
   }
 
   for (auto& key : TensorPipeChannelRegistry()->Keys()) {
+    int64_t priority = -1;
+    if (opts_.channels.has_value()) {
+      auto iter =
+          std::find(opts_.channels->begin(), opts_.channels->end(), key);
+      if (iter == opts_.channels->end()) {
+        continue;
```

Comment:
This means absent transport/channels are not registered (killed)?

Reply:
Exactly. This is in fact the main purpose of these fields. If you prefer I can do it the other way around, and have these fields be "blacklists" for backends, rather than "whitelists".